src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateClones = [ { ...state } ];
+  const stateClone = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        for (const key in action.extraData) {
+          stateClone[key] = action.extraData[key];
+        }
+        break;
+
+      case 'removeProperties':
+        for (const keyToRemove of action.keysToRemove) {
+          delete stateClone[keyToRemove];
+        }
+        break;
+
+      case 'clear':
+        for (const key in stateClone) {
+          delete stateClone[key];
+        }
+        break;
+    }
+
+    stateClones.push({ ...stateClone });
+  }
+
+  return stateClones.slice(1);
```

Comment:
it's better to use spread operator for making a copy 
```suggestion
  return {...stateClones}
```